aconcli/service/aconclient_http.go

```diff
@@ -226,6 +253,130 @@ func multipartManifestField(w *multipart.Writer, fieldname, filename string) err
+	}
+	if _, err = processReponse(resp); err != nil {
+		return fmt.Errorf(clientProcRespErrFmt, "Logout", err)
+	}
+
+	if err := RemoveAuthToken(uid, vmid); err != nil {
+		return fmt.Errorf("fail to log out: %v", err)
+	}
+	return nil
+}
+
+func (c *AconClientHttp) Login(uid string, vmid string) error {
+	clientId := os.Getenv("ATD_CLIENT_ID")
+	if clientId == "" {
+		return fmt.Errorf("failed to get env variable ATD_CLIENT_ID for authentication")
+	}
+	clientSec := os.Getenv("ATD_CLIENT_SECRET")
+	if clientSec == "" {
+		return fmt.Errorf("failed to get env variable ATD_CLIENT_SECRET for authentication")
+	}
+	resp, err := http.Get(CONFIG_URL_GOOGLE)
+	if err != nil {
+		return fmt.Errorf("failed to get openid config: %v", err)
+	}
+	configJson, err := io.ReadAll(resp.Body)
+	resp.Body.Close()
+	if err != nil {
+		return fmt.Errorf("failed to read openid config: %v", err)
+	}
+	var config OpenidConfig
+	if err := json.Unmarshal(configJson, &config); err != nil {
+		return fmt.Errorf("failed to extract openid config: %v", err)
+	}
+
+	authEndpoint := config.DeviceAuthEndpoint
+	resp, err = http.PostForm(authEndpoint,
+		url.Values{
+			"client_id": {clientId},
+			"scope":     {"email"}})
+	if err != nil {
+		return fmt.Errorf("failed to get response from auth server (%s): %v",
+			authEndpoint, err)
+	}
+	defer resp.Body.Close()
+	authCodeJson, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read response body from auth server (%s): %v",
+			authEndpoint, err)
+	}
+	var auth AuthCode
+	if err := json.Unmarshal(authCodeJson, &auth); err != nil {
+		return fmt.Errorf("failed to extract auth code info: %v", err)
+	}
+	fmt.Printf("\nPlease visit the following URL via a web browser:\n\n%s\n\n"+
+		"and fill in the following code:\n\n%s\n\n", auth.VerificationURL, auth.UserCode)
+
+	resp, err = c.client.PostForm(c.makeURL(endpointLogin),
+		url.Values{
+			"client_id":     {clientId},
+			"client_secret": {clientSec},
+			"device_code":   {auth.DeviceCode},
+			"expires_in":    {strconv.FormatUint(uint64(auth.Expiration), 10)},
+			"interval":      {strconv.FormatUint(uint64(auth.Interval), 10)}})
+	if err != nil {
+		return fmt.Errorf("failed to get access token: %v", err)
+	}
+	defer resp.Body.Close()
+	keydata, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read response for access token: %v", err)
+	}
+	var key string
+	if err := json.Unmarshal(keydata, &key); err != nil {
+		return fmt.Errorf("failed to parse access token from response: %v", err)
+	}
+	if err := UpdateAuthToken(uid, map[string]string{vmid: key}); err != nil {
+		return fmt.Errorf("failed to update access token: %v", err)
+	}
+	c.sessionkey = key
+	return nil
+}
+
+func (c *AconClientHttp) fetchSessionKey() error {
+	if len(c.sessionkey) > 0 {
+		return nil
+	}
+	user, err := user.Current()
+	if err != nil {
+		return fmt.Errorf("failed to get current user: %v", err)
+	}
+	key, err := GetAuthToken(user.Uid, c.host)
```

Comment:
What if an ACON TD shuts down without `aconcli` knowing? Is the AuthToken table cleaned up regularly?

Reply:
Added the code to cleanup expired auth token entries